runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -2164,6 +2228,151 @@ J9::ValuePropagation::isArrayCompTypeValueType(TR::VPConstraint *arrayConstraint
+                                                            fieldEntry._isFinal,
+                                                            fieldEntry._fieldname,
+                                                            fieldEntry._typeSignature
+                                                            );
+
+      if (trace())
+         {
+         traceMsg(comp(), "%s Changing n%dn from %s to %s fieldEntry[0] fieldName %s fieldSig %s type %d offset %d\n", __FUNCTION__,
+            callNode->getGlobalIndex(), isObjectEqualityCompare ? "<objectEqualityComparison>" : "<objectInequalityComparison>",
+            comp()->getDebug()->getName(cmpOpCode), fieldEntry._fieldname, fieldEntry._typeSignature, dataType.getDataType(), fieldEntry._offset);
+         traceMsg(comp(),"    %s loadFieldSymRef %p %s \n", comp()->getDebug()->getName(loadOpCode), loadFieldSymRef, comp()->getDebug()->getName(loadFieldSymRef));
+         }
+
+      TR::Node *loadLhsFieldNode = TR::Node::createWithSymRef(loadOpCode, 1, 1, lhsNode, loadFieldSymRef);
+      TR::Node *loadRhsFieldNode = TR::Node::createWithSymRef(loadOpCode, 1, 1, rhsNode, loadFieldSymRef);
+
+      TR::Node::recreateWithoutProperties(callNode, cmpOpCode, 2, loadLhsFieldNode, loadRhsFieldNode);
+      }
+   else
+      {
+      /* Before transformation:
+       * n40n   treetop
+       * n39n     icall  <objectInequalityComparison>
+       * n30n       aloadi  ACMPWideFieldsPrimitive$Holder._wfp QWideFieldsPrimitive;
+       * n28n         ...
+       * n36n       aload  ACMPWideFieldsPrimitive.VALUE1 QWideFieldsPrimitive;
+       *
+       * After transformation:
+       * n40n   treetop
+       * n39n     arraycmp  <arraycmp>
+       * n100n      aladd
+       * n30n         aloadi  ACMPWideFieldsPrimitive$Holder._wfp QWideFieldsPrimitive;
+       * n28n           ...
+       * n99n         lconst 4
+       * n101n      aladd
+       * n36n         aload  ACMPWideFieldsPrimitive.VALUE1 QWideFieldsPrimitive;
+       * n99n         ==>lconst 4
+       * n98n       iconst 12
+       */
+      int32_t totalFieldSize = 0;
+      for (size_t idx = 0; idx < fieldCount; idx++)
+         {
+         const TR::TypeLayoutEntry &fieldEntry = fieldTypeLayout->entry(idx);
+         TR::DataType dataType = fieldEntry._datatype;
+
+         totalFieldSize += TR::DataType::getSize(fieldEntry._datatype);
+         }
+
+      TR::Node *totalFieldSizeNode = TR::Node::iconst(callNode, totalFieldSize);
+
+      TR::Node * lhsOffsetNode = NULL;
+      TR::Node * rhsOffsetNode = NULL;
+      TR::Node * offsetNode = NULL;
+
+      // Skip the object header
+      if (comp()->target().is64Bit())
+         {
+         offsetNode = TR::Node::create(callNode, TR::lconst, 0, 0);
+         offsetNode->setLongInt(TR::Compiler->om.objectHeaderSizeInBytes());
+         lhsOffsetNode = TR::Node::create(TR::aladd, 2, lhsNode, offsetNode);
+         rhsOffsetNode = TR::Node::create(TR::aladd, 2, rhsNode, offsetNode);
```

Comment:
Does this need to worry about compressed references?

Reply:
It looks to me `J9::ObjectModel::objectHeaderSizeInBytes` considers compressed reference case:

https://github.com/eclipse-openj9/openj9/blob/8c1840d808785e26daf7eac6096d5c6be17f1c3e/runtime/compiler/env/J9ObjectModel.cpp#L540-L543

Reply:
I was actually worried about the `aladd` rather than the header size.  Does it need to be placed under a `compressedrefs` anchor when a compressed reference is used?  I must confess that I'm really fuzzy on when and how the `compressedrefs` anchor is supposed to be used.

Reply:
I looked at how `aladd` is used in other parts of the code and it doesn't look like it's placed under a `compressedrefs` 
such as in `TR_UnsafeFastPath::perform`. I do think I need to add `setIsInternalPointer(true)`. https://github.com/eclipse-openj9/openj9/blob/afa5334f999126659dcf3a3385fd1ce5da067b6c/runtime/compiler/optimizer/UnsafeFastPath.cpp#L811-L819

Reply:
>I was actually worried about the aladd rather than the header size. Does it need to be placed under a compressedrefs anchor when a compressed reference is used?

@jdmpapin Could you shed some light on this one? I haven't found an existing example on using `aladd` that is put under `compressedrefs`

Reply:
@a7ehuo Annabelle, In taking another look at [IL generation for field references](https://github.com/eclipse-openj9/openj9/blob/3db1d838ccc0ed44b408225e32dd696cc44a7aac/runtime/compiler/ilgen/Walker.cpp#L5150-L5168) and how [Value Propagation transforms the `<jitLoadFlatttenableArrayElement>` helper](https://github.com/eclipse-openj9/openj9/blob/3db1d838ccc0ed44b408225e32dd696cc44a7aac/runtime/compiler/optimizer/J9ValuePropagation.cpp#L2314-L2326), I see that the addresses involved in the comparison will already be placed under `compressedRefs` anchors when required.  So the situation I was worried about doesn't exist.

You can mark this as resolved.

Reply:
Thanks @hzongaro! I checked the test logs which confirm that instance field load node is anchored under `compressedRefs`. The `compressedRefs` of the children of `aladd` here are taken care of when they are generated. 

```
n2254n    compressedRefs                                                                      [0x7f66f721b040] bci=[-1,17,28] rc=0 vc=0 vn=- li=- udi=- nc=2
n28n        aloadi  <array-shadow>[#232  Shadow] [flags 0x80000607 0x0 ]                      [0x7f66f7005000] bci=[-1,17,28] rc=3 vc=406 vn=- li=- udi=- nc=1
n2252n        aladd (internalPtr sharedMemory )                                               [0x7f66f721afa0] bci=[-1,13,28] rc=1 vc=0 vn=- li=- udi=- nc=2 flg=0x8000
...
...
n31n      NULLCHK on n28n [#32]                                                               [0x7f66f70050f0] bci=[-1,18,28] rc=0 vc=406 vn=- li=- udi=- nc=1
n30n        aloadi  ACMPWideFieldsPrimitiveStaticInstanceAlwaysEqual$Holder._wfp QWideFieldsPrimitive;[#361  Shadow +8] [flags 0x607 0x0 ]  [0x7f66f70050a0] bci=[-1,18,28] rc=4 vc=406 vn=- li=- udi=- nc=1
n28n          ==>aloadi
n33n      compressedRefs                                                                      [0x7f66f7005190] bci=[-1,18,28] rc=0 vc=406 vn=- li=- udi=- nc=2
n30n        ==>aloadi
n32n        lconst 0 (highWordZero X==0 X>=0 X<=0 )                                           [0x7f66f7005140] bci=[-1,18,28] rc=1 vc=406 vn=- li=- udi=- nc=0 flg=0x4302
...
n38n      treetop                                                                             [0x7f66f7005320] bci=[-1,27,28] rc=0 vc=406 vn=- li=- udi=- nc=1
n36n        aload  ACMPWideFieldsPrimitiveStaticInstanceAlwaysEqual.VALUE1 QWideFieldsPrimitive;[#363  Static] [flags 0x80307 0x0 ]  [0x7f66f7005280] bci=[-1,24,28] rc=2 vc=406 vn=- li=- udi=- nc=0
n40n      treetop                                                                             [0x7f66f70053c0] bci=[-1,30,28] rc=0 vc=406 vn=- li=- udi=- nc=1
n39n        arraycmp  <arraycmp>[#221  helper Method] [flags 0x400 0x0 ] ()                   [0x7f66f7005370] bci=[-1,30,28] rc=2 vc=406 vn=- li=- udi=- nc=3 flg=0x20
n2243n        aladd                                                                           [0x7f66f721acd0] bci=[-1,18,28] rc=1 vc=0 vn=- li=- udi=- nc=2
n30n            ==>aloadi
n2242n          lconst 4                                                                      [0x7f66f721ac80] bci=[-1,30,28] rc=2 vc=0 vn=- li=- udi=- nc=0
n2244n        aladd                                                                           [0x7f66f721ad20] bci=[-1,24,28] rc=1 vc=0 vn=- li=- udi=- nc=2
n36n            ==>aload
n2242n          ==>lconst 4
n2241n        iconst 12
```